Changes in module 'nf-core/seqkit/grep'
--- modules/nf-core/seqkit/grep/main.nf
+++ modules/nf-core/seqkit/grep/main.nf
@@ -13,7 +13,7 @@
     path pattern
 
     output:
-    tuple val(meta), path("*.{fa,fq}.gz")  , emit: filter
+    tuple val(meta), path("*.{fa,fq}")  , emit: filter
     path "versions.yml"                    , emit: versions
 
     when:
@@ -33,7 +33,7 @@
         --threads $task.cpus \\
         ${pattern_file} \\
         ${sequence} \\
-        -o ${prefix}.${suffix}.gz \\
+        -o ${prefix}.${suffix} \\
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
