Changes in module 'nf-core/bioawk'
--- modules/nf-core/bioawk/main.nf
+++ modules/nf-core/bioawk/main.nf
@@ -11,7 +11,7 @@
     tuple val(meta), path(input)
 
     output:
-    tuple val(meta), path("*.gz"), emit: output
+    tuple val(meta), path("*.fasta"), emit: output
     path "versions.yml"             , emit: versions
 
     when:
@@ -26,9 +26,7 @@
     bioawk \\
         $args \\
         $input \\
-        > ${prefix}
-
-    gzip ${prefix}
+        > ${prefix}.fasta
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
